libbpf-tools/readahead.bpf.c

```diff
@@ -0,0 +1,87 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Wenbo Zhang
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+#include "readahead.h"
+#include "bits.bpf.h"
+
+#define MAX_ENTRIES 10240
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u32);
+	__type(value, u8);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} in_readahead SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct page *);
+	__type(value, u64);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} birth SEC(".maps");
+
+static struct hist hist;
+
+SEC("fentry/__do_page_cache_readahead")
+int BPF_PROG(fentry____do_page_cache_readahead)
+{
+	u32 pid = bpf_get_current_pid_tgid();
+
+	bpf_map_update_elem(&in_readahead, &pid, &pid, 0);
```

Comment:
value type is `u8`, yet you are trying to put pid there?

Reply:
Yes, this is my intention. I only care about the existence of the key in the map, but not the val value. So use u8 to save memory.

Reply:
hm... so the value doesn't matter, ok. But it's not clear, so I wouldn't do it, I'd rather do `u64 one = 1;` and use that as a value. Also, value size internally in kernel is rounded up to 8 bytes anyways, so you are not saving anything, really.

Reply:
Got it, thanks! I'll change it to make it clear.